pkg/core/policy/dataplane_matcher.go

```diff
@@ -4,16 +4,19 @@ import (
 	"sort"
 
 	mesh_proto "github.com/Kong/kuma/api/mesh/v1alpha1"
-	"github.com/Kong/kuma/pkg/core/resources/apis/mesh"
 )
 
-// SelectDataplanePolicy given a Dataplane definition and a list of DataplanePolicy returns the "best matching" DataplanePolicy.
-// A DataplanePolicy is considered a match if one of the inbound interfaces of a Dataplane or tag section on Gateway Dataplane has all tags of DataplanePolicy's selector.
+type MatchingTags interface {
+	MatchTags(selector mesh_proto.TagSelector) bool
+}
+
+// SelectDataplanePolicy given a something that could be matched (mesh_proto.Dataplane,
+// mesh_proto.Dataplane_Networking_Inbound, etc.) and a list of DataplanePolicy returns the "best matching" DataplanePolicy.
 // Every matching DataplanePolicy gets a rank (score) defined as a maximum number of tags in a matching selector.
 // DataplanePolicy with an empty list of selectors is considered a match with a rank (score) of 0.
 // DataplanePolicy with an empty selector (one that has no tags) is considered a match with a rank (score) of 0.
 // In case if there are multiple DataplanePolicies with the same rank (score), the policy created last is chosen.
-func SelectDataplanePolicy(dataplane *mesh.DataplaneResource, policies []DataplanePolicy) DataplanePolicy {
+func SelectDataplanePolicy(matching MatchingTags, policies []DataplanePolicy) DataplanePolicy {
```

Comment:
This is clever solution, but can we take a step back?

The solution with reusing DataplanePolicy for matching ConnectionPolicy with adapter is DRY, but I must say this is a little bit hard to follow. This is one of the most important logic in our code, so I want it to be as understandable as possible.

I'd rather have simple understandable interfaces with clear input. I prefer to have a little bit of copying instead of using wrong abstraction (https://www.sandimetz.com/blog/2016/1/20/the-wrong-abstraction)

Additionally, the code from FaultInjectionMatcher should be in `policy` since it will be soon used for TrafficPermission and also sooner or later on TrafficLogs, so `FaultInjectionMatcher` should be as thin as possible.

I experimented on branch and here is the result https://github.com/Kong/kuma/commit/bc37646d9a9943f0ec9e870267c5142c99d18c64
we don't need te full logic from DataplaneMatcher because with ConnectionPolicies we enforce that there are sources and destinations.

What do you think?